packages/material-ui/src/ExpansionPanel/ExpansionPanel.js

```diff
@@ -78,93 +78,87 @@ export const styles = theme => {
   };
 };
 
-class ExpansionPanel extends React.Component {
-  constructor(props) {
-    super();
-    this.isControlled = props.expanded != null;
-    this.state = {};
-    if (!this.isControlled) {
-      // not controlled, use internal state
-      this.state.expanded = props.defaultExpanded !== undefined ? props.defaultExpanded : false;
+function useIsControlled(value) {
+  const { current: isControlled } = React.useRef(value != null);
+  return isControlled;
+}
+
+function ExpansionPanel(props) {
+  const {
+    children: childrenProp,
+    classes,
+    className: classNameProp,
+    CollapseProps,
+    defaultExpanded,
+    disabled,
+    expanded: expandedProp,
+    onChange,
+    square,
+    ...other
+  } = props;
+
+  const isControlled = useIsControlled(expandedProp);
+  const [expandedState, setExpandedState] = React.useState(defaultExpanded);
+
+  const expanded = isControlled ? expandedProp : expandedState;
+
+  function handleChange(event) {
+    if (!isControlled) {
+      setExpandedState(isExpanded => !isExpanded);
+    }
+    if (onChange) {
+      onChange(event, !expanded);
     }
   }
 
-  handleChange = event => {
-    const expanded = this.isControlled ? this.props.expanded : this.state.expanded;
+  let summary = null;
 
-    if (!this.isControlled) {
-      this.setState({ expanded: !expanded });
+  const children = React.Children.map(childrenProp, child => {
+    if (!React.isValidElement(child)) {
+      return null;
     }
 
-    if (this.props.onChange) {
-      this.props.onChange(event, !expanded);
+    warning(
+      child.type !== React.Fragment,
+      [
+        "Material-UI: the ExpansionPanel component doesn't accept a Fragment as a child.",
+        'Consider providing an array instead.',
+      ].join('\n'),
+    );
+
+    if (isMuiElement(child, ['ExpansionPanelSummary'])) {
```

Comment:
How are you going to solve this problem? Though, I don't see the value of using the context if we still have to expose `.muiName = 'ExpansionPanelSummary'`.

Reply:
I was a bit premature with that statement. As soon as I looked at the code I realized we wouldn't gain as much as I initially thought.

What context would allow though is reaching into the `ExpansionPanelSummary`, grabbing the html ID, expose this in the context and then the content can get a `aria-labelledby`.

Context is definitely more powerful here. I need to finish up the implementation to fully understand what features this enables.

Reply:
I'm happy that you are mentioning a11y. So far, we have had a hard time abstracting the accessibility label logic. It often requires the introducing of a global index structure or the usage of the context to propagate the changes down the tree. Our tradeoff so far has always been to defer the logic to the developers, saving, what we thought bundle size.

Reply:
Context is currently my preferred solution to these problems. I like neither the `SomePassedComponent` prop + `SomePassedComponentProps` prop pattern nor `cloneElement`. It feels like black magic from the outside. 

The Context + `useSomeExposedPartFromTheContext` custom hook pattern is way more predictable. While it might be more bundle size it resembles more what we're actually trying to achieve. The previous patterns only tried to emulate some context. Now we actually have some. Without the downsides of possibly defeating memoized components or props collision.

And with the new `react-devtools` debugging should be easier i.e. you actually see where the props are coming from.